flask/app.py

```diff
@@ -1982,6 +1997,10 @@ def wsgi_app(self, environ, start_response):
                                a list of headers and an optional
                                exception context to start the response
         """
+        with self._before_request_lock:
+            if not self._got_first_request and self._should_add_static_route:
+                self._add_static_route()
+                self._should_add_static_route = False
```

Comment:
I was concerned that only using the `_got_first_request` guard inside the new, earlier `with self._before_request_lock` block I added\*, could introduce a race condition:

If the first two requests come in at almost the same time, the first one enters the first `with ...lock` block and adds the static route, then the second one enters that first `with ...lock` block and *re*-adds the static route, *before* the first  request enters the second `with ...lock` block and finally flips the `_got_first_request` guard to True. So I added the additional guard for whether we've added the static route yet.

But after your comment, and upon taking another look, I'm wondering whether this race condition can actually happen, or whether adding the same route twice is not a big deal in any case. If not, I've updated the logic to match your suggestions in 8ba6f35. If that's sound, this should now be ready to merge, otherwise I'm happy to work through this further.

Thanks for taking another look.

\* which comes before the existing `with self._before_request_lock` block in `try_trigger_before_first_request_functions()` (called by `full_dispatch_request()`, which is called by `wsgi_app()` a few lines after the logic I added), and which is the block that finally sets the `_got_first_request` guard to True